/* Copyright (c) 2010-2011 Stanford University
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package RAMCloud.ProtoBuf;

// The information appended to LogCabin by the Coordinator
// to store the information about each table.
message TableInformation {

  message TabletInfo {
    /// The smallest hash value for a key that is in this tablet.
    required uint64 start_key_hash = 1;

    /// The largest hash value for a key that is in this tablet.
    required uint64 end_key_hash = 2;

    /// The ID of the master owning this tablet.
    required fixed64 master_id = 3;

    /// Id of the log head segment belonging to master_id
    /// when it was assigned to the server. Any earlier segments
    /// cannot contain data belonging to this tablet.
    required uint64 ctime_log_head_id = 8;

    /// Offset in the log head segment belonging to master_id
    /// when it was assigned to the server. Any objects appearing
    /// earlier in that segment cannot contain data belonging to this tablet.
    required uint32 ctime_log_head_offset = 9;
  }

  /// Entry type for this LogCabin entry.
  required string entry_type = 1;

  /// The name of the table.
  required string name = 2;

  /// The id of the table.
  required uint64 table_id = 3;

  /// Number of servers across which this table is split.
  required uint32 server_span = 4;

  /// The tablets.
  repeated TabletInfo tablet_info = 5;
}
